### ISSUE TITLE
Use OSGI environment facility to load a native library

### ISSUE THREAD

**Issue by bjmi** [luben/zstd-jni#130]
OSGI provides a mechanism to load multiple native libraries for different architectures and operating systems in a single bundle.
For this
1.  Add `Bundle-NativeCode` header to _MANIFEST.MF_
```
Bundle-NativeCode: aix/ppc64/libzstd-jni.so;osname=AIX;processor=ppc64,
 darwin/x86_64/libzstd-jni.dylib;osname=MacOS;osname=MacOSX;processor=x86_64,
 freebsd/amd64/libzstd-jni.so;osname=FreeBSD;processor=amd64,
 freebsd/i386/libzstd-jni.so;osname=FreeBSD;processor=i386,
 linux/aarch64/libzstd-jni.so;osname=Linux;processor=aarch64,
 linux/amd64/libzstd-jni.so;osname=Linux;processor=amd64,
 linux/arm/libzstd-jni.so;osname=Linux;processor=arm,
 linux/i386/libzstd-jni.so;osname=Linux;processor=i386,
 linux/mips64/libzstd-jni.so;osname=Linux;processor=Mips64,
 linux/ppc64/libzstd-jni.so;osname=Linux;processor=ppc64,
 linux/ppc64le/libzstd-jni.so;osname=Linux;processor=ppc64le,
 win/amd64/libzstd-jni.dll;osname=Win32;processor=amd64,
 win/x86/libzstd-jni.dll;osname=Win32;processor=x86,
 *
```
(Couldn't find values for _mips64_ and _aarch64_ in the reference)

2. Load native library in OSGI environment without a copy in _temp_ folder
```
try {
    Class.forName("org.osgi.framework.BundleEvent"); // Simple OSGI env. check
    System.loadLibrary(libname);
    loaded = true;
    return;
} catch (final Exception e) {
   // Ignore and try other methods
}
```
OSGI runtime takes _Bundle-NativeCode_ entries into account and loads the appropriate library when `System.loadLibrary(String)` is called.
`Eclipse-BundleShape: dir` introduced by #129 can be dropped again.

### References
* [Bundle-NativeCode Header](https://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module-loading.native.code.libraries)
* [OS Names](https://docs.osgi.org/specification/osgi.core/7.0.0/framework.lifecycle.html#framework.module.os-names)
* [Processor Names](https://docs.osgi.org/specification/osgi.core/7.0.0/framework.lifecycle.html#framework.module.processor-names)
* Usage example in [JNA](https://github.com/java-native-access/jna/blob/841ae46920adb4d0c0522eb5ad3c87a5d9c37756/build.xml#L458)

**Comment by bjmi** [comment]
Supplement: `Class.forName("org.osgi.framework.BundleEvent")` check only works if _MANIFEST.MF_ contains
`Import-Package: org.osgi.framework;resolution:=optional`

**Comment by luben** [comment]
Ok, I will try that.

**Comment by luben** [comment]
Hi, I am on a vacation with spotty internet access. I will look into it in 2-3 weeks

**Comment by luben** [comment]
Hi, can you check the last commit - it looks like it's producing the right manifest but I have not tested loading directly from the bundle.

**Comment by bjmi** [comment]
I don't know _sbt_ well. Running `sbt compile package` in WSL2 on my Windows machine produces a file _zstd-jni-1.4.5-7.jar_ with this _MANIFEST.MF_
```
Manifest-Version: 1.0
Implementation-Title: zstd-jni
Automatic-Module-Name: com.github.luben.zstd_jni
Implementation-Version: 1.4.5-7
Bundle-NativeCode: aix/ppc64/libzstd-jni.so;osname=AIX;processor=ppc64
 ,
darwin/x86_64/libzstd-jni.dylib;osname=MacOS;osname=MacOSX;processo
 r=x86_64,
freebsd/amd64/libzstd-jni.so;osname=FreeBSD;processor=amd64
 ,
freebsd/i386/libzstd-jni.so;osname=FreeBSD;processor=i386,
linux/aa
 rch64/libzstd-jni.so;osname=Linux;processor=aarch64,
linux/amd64/libz
 std-jni.so;osname=Linux;processor=amd64,
linux/arm/libzstd-jni.so;osn
 ame=Linux;processor=arm,
linux/i386/libzstd-jni.so;osname=Linux;proce
 ssor=i386,
linux/mips64/libzstd-jni.so;osname=Linux;processor=mips64,
 
linux/ppc64/libzstd-jni.so;osname=Linux;processor=ppc64,
linux/ppc64
 le/libzstd-jni.so;osname=Linux;processor=ppc64le,
win/amd64/libzstd-j
 ni.dll;osname=Win32;processor=amd64,
win/x86/libzstd-jni.dll;osname=W
 in32;processor=x86
Specification-Vendor: com.github.luben
Specification-Title: zstd-jni
Implementation-Vendor-Id: com.github.luben
Specification-Version: 1.4.5-7
Implementation-Vendor: com.github.luben
```
The _Export-Package_, _Import-Package_ etc. are missing. Additionally the line breaks for `Bundle-NativeCode` look strange too. Previously everything was wrapped after 70 characters. Can you confirm this too or can you provide the _MANIFEST.MF_ produced by CI?

Thanks in advance

**Comment by luben** [comment]
Hi,

Try with `sbt compile publishLocal` and look for the packages in `~/.ivy/local`. Here is what I get for the manifest:

```
Manifest-Version: 1.0
Automatic-Module-Name: com.github.luben.zstd_jni
Bnd-LastModified: 1599561390599
Bundle-Description: JNI bindings for Zstd native library that provides
  fast and high compression lossless algorithm for Java and all JVM la
 nguages.
Bundle-License: https://opensource.org/licenses/BSD-2-Clause;descripti
 on=BSD 2-Clause License
Bundle-ManifestVersion: 2
Bundle-Name: zstd-jni
Bundle-NativeCode: aix/ppc64/libzstd-jni.so;osname=AIX;processor=ppc64
 , darwin/x86_64/libzstd-jni.dylib;osname=MacOS;osname=MacOSX;processo
 r=x86_64, freebsd/amd64/libzstd-jni.so;osname=FreeBSD;processor=amd64
 , freebsd/i386/libzstd-jni.so;osname=FreeBSD;processor=i386, linux/aa
 rch64/libzstd-jni.so;osname=Linux;processor=aarch64, linux/amd64/libz
 std-jni.so;osname=Linux;processor=amd64, linux/arm/libzstd-jni.so;osn
 ame=Linux;processor=arm, linux/i386/libzstd-jni.so;osname=Linux;proce
 ssor=i386, linux/mips64/libzstd-jni.so;osname=Linux;processor=mips64,
  linux/ppc64/libzstd-jni.so;osname=Linux;processor=ppc64, linux/ppc64
 le/libzstd-jni.so;osname=Linux;processor=ppc64le, win/amd64/libzstd-j
 ni.dll;osname=Win32;processor=amd64, win/x86/libzstd-jni.dll;osname=W
 in32;processor=x86
Bundle-SymbolicName: com.github.luben.zstd-jni
Bundle-Vendor: com.github.luben
Bundle-Version: 1.4.5.7
Created-By: 1.8.0_265 (Private Build)
Export-Package: com.github.luben.zstd;version="1.4.5.7",com.github.lub
 en.zstd.util;version="1.4.5.7"
Implementation-Title: zstd-jni
Implementation-Vendor: com.github.luben
Implementation-Vendor-Id: com.github.luben
Implementation-Version: 1.4.5-7
Import-Package: org.osgi.framework;resolution:=optional
Private-Package: include,linux.amd64
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Specification-Title: zstd-jni
Specification-Vendor: com.github.luben
Specification-Version: 1.4.5-7
Tool: Bnd-4.0.0.201805111645
```

note, it includes only the linux/amd64 binary

**Comment by bjmi** [comment]
I tested it on Windows 10 64 bit with Eclipse 4.14 (2019-12) and it works.

The dll is extracted to `C:\dev\runtime-workspace\configuration\org.eclipse.osgi\255\0\.cp\win\amd64\libzstd-jni.dll` _once_ and is reused after restart. No files like _libzstd-jni12317205200951817934.dll_ were created in temp folder.

Thank you.

**Comment by luben** [comment]
Great, I will release it in a few weeks.

**Comment by luben** [comment]
Released in 1.4.5-7
